rechunker/executors/beam.py

```diff
@@ -0,0 +1,94 @@
+from functools import partial
+import itertools
+import math
+import uuid
+from typing import Any, Iterable, Optional, Mapping, Tuple
+
+import apache_beam as beam
+
+from rechunker.types import CopySpec, StagedCopySpec, Executor
+
+
+class BeamExecutor(Executor[beam.PTransform]):
+    """An execution engine based on Apache Beam.
+
+    Supports copying between any arrays that implement ``__getitem__`` and
+    ``__setitem__`` for tuples of ``slice`` objects. Array must also be
+    serializable by Beam (i.e., with pickle).
+
+    Execution plans for BeamExecutor are beam.PTransform objects.
+    """
+
+    # TODO: explore adding an option to do rechunking with Beam groupby
+    # operations instead of explicitly writing intermediate arrays to disk.
+
+    def prepare_plan(self, specs: Iterable[StagedCopySpec]) -> beam.PTransform:
+        return "Rechunker" >> _Rechunker(specs)
+
+    def execute_plan(self, plan: beam.PTransform, **kwargs):
+        with beam.Pipeline(**kwargs) as pipeline:
+            pipeline | plan
+
+
+class _Rechunker(beam.PTransform):
+    def __init__(self, specs: Iterable[StagedCopySpec]):
+        super().__init__()
+        self.specs = tuple(specs)
+
+    def expand(self, pcoll):
+        max_depth = max(len(spec.stages) for spec in self.specs)
+        specs_map = {uuid.uuid1().hex: spec for spec in self.specs}
+
+        pcoll = pcoll | "Create" >> beam.Create([(k, None) for k in specs_map])
+        for stage in range(max_depth):
+            specs_by_target = {
+                k: v.stages[stage] if stage < len(v.stages) else None
+                for k, v in specs_map.items()
+            }
+            pcoll = pcoll | f"Stage{stage}" >> _Stage(specs_by_target)
+        return pcoll
+
+
+class _Stage(beam.PTransform):
+    def __init__(self, specs_by_target: Mapping[str, CopySpec]):
+        super().__init__()
+        self.specs_by_target = specs_by_target
+
+    def expand(self, pcoll):
+        start_fn = partial(_start_stage, self.specs_by_target)
+        return (
+            pcoll
+            | "Start" >> beam.FlatMapTuple(start_fn)
+            | "CreateTasks" >> beam.FlatMapTuple(_copy_tasks)
+            # prevent undesirable fusion
+            # https://stackoverflow.com/a/54131856/809705
+            | "Reshuffle" >> beam.Reshuffle()
+            | "CopyChunks" >> beam.MapTuple(_copy_chunk)
+            | "Finish" >> beam.GroupByKey()
```

Comment:
Yes, I think this is important. Otherwise there's nothing explicitly keeping track of the dependency order between stages and I believe Beam would be free (in principle) to run the stages in any order.

This is actually the entire reason why I'm explicitly passing around `target_id` in my pipeline at all.

Looking at this more carefully, I realize that I can at least get rid of all the `None` objects by using `beam.Distinct()` here. So I've done that instead now.

Reply:
Thanks for the explanation. I think understand the code now. It’s very clever. It builds one static beam “graph” that you can push all the StagedCopySpecs (e.g. variables to reshape) into.  This graph must have as many stages as the longest StagedCopySpec it wants to process. You pass the “target_id” and call `beam.Distinct` in as a way to synchronize the storage operations after each stage.

For some reason, beam pipelines remind me of an elaborate mail sorting machine. It’s tricky to install/setup, but then it just flies through the individual envelopes.